source/Higgs/UntruncatedAtMostTwo.lagda

```diff
@@ -0,0 +1,244 @@
+J. A. Carr 8 July 2025.
+
+The untruncated form of the at-most-two elements is equivalent
+to the statement that Aut Ω has exactly 1 or 2 elements
+(and hence every element is either identity or negation)
+
+\begin{code}
+
+{-# OPTIONS --safe --without-K #-}
+
+open import MLTT.Spartan
+open import MLTT.Plus-Properties
+open import UF.DiscreteAndSeparated
+open import TypeTopology.SigmaDiscreteAndTotallySeparated
+open import UF.Equiv hiding (_≅_)
+open import UF.FunExt
+open import UF.Logic
+open import UF.PropTrunc
+open import UF.Subsingletons
+open import UF.SubtypeClassifier hiding (Ω)
+open import Fin.Type
+open import Fin.Topology
+open import Fin.Pigeonhole
+
+module Higgs.UntruncatedAtMostTwo
+        {𝓤 : Universe}
+        (fe : Fun-Ext)
+        (pe : propext 𝓤)
+        (pt : propositional-truncations-exist)
+       where
+
+open import Higgs.Rigidity fe pe
+open import Higgs.InvolutionTheorem fe pe
+open import Higgs.AutomorphismsOfOmega fe pe
+open import Higgs.AutomorphismsOfOmegaWEM fe pe pt
+
+open Conjunction
+open Implication fe
+open Universal fe
+
+\end{code}
+
+Assuming function extensionality, having an untruncated pigeonhole principle
+reflects discreteness to the codomain.
+
+The core method is to note that the pidgeonhole function must respect equality
+of functions, so we produce a pair of functions, where equality of their results
+holds if and only if the two elements of the codomain are equal.
+
+\begin{code}
+
+constantly : {𝓤 𝓥 : Universe}
+           → {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+           → Y → X → Y
+constantly const-y _ = const-y
+
+almost-constantly : {𝓤 𝓥 : Universe}
+                  → {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+                  → is-discrete X
+                  → X → Y → Y → X → Y
+almost-constantly X-discrete pt-x pt-y const-y x' =
+  cases
+   (λ _ → pt-y) (λ _ → const-y)
+   (X-discrete x' pt-x)
+
+almost-constantly-is-constant
+ : {𝓤 𝓥 : Universe}
+ → {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+ → (X-discrete : is-discrete X)
+ → (x : X)
+ → (y y' : Y)
+ → y ＝ y'
+ → constantly y ＝ almost-constantly X-discrete x y' y
+almost-constantly-is-constant {_} {_} {X} {Y} X-discrete x y _ refl = dfunext fe I
+ where
+  I : (x' : X)
+    → constantly y x' ＝ almost-constantly X-discrete x y y x'
+  -- possibly rewrite this with dep-cases?
+  -- it's more obvious and spartan but much more awkward
+  I x' with X-discrete x' x
```

Comment:
We don't use `with` in TypeTopology. There are a few `with`s left, but the intension is to get rid of all of them. 

Reasons (among others):
(1) It is difficult to prove things about things defined with `with`.
(2) It has a rather unclear semantics, and we try to use a fragment of Agda that amounts to MLTT.

Reply:
Use e.g. `Cases` or `dep-Cases` instead, or write a helper function (I often do the latter).

Reply:
I fixed this, although it made the proofs significantly more verbose